File: internal/lifecycle/restore.go
package lifecycle

import (
	"context"
	"fmt"

	"github.com/abbyfluoroethane/bento/internal/types"
)

// Restore brings the host back to the last recorded desired state after a
// reboot (SPEC 11.2):
//
//  1. Read the observed state of every domain from libvirt.
//  2. Select every instance with desired running and observed stopped.
//  3. Start those instances in batches (operator setting, default 4).
//  4. Wait for each batch to reach running before the next batch.
//
// An instance a user left running comes back running; one a user stopped
// stays stopped. Progress goes to the logger; a user who connects during
// the restore sees "starting", not an error.
func (m *Manager) Restore(ctx context.Context) error {
	domains, err := m.hyp.List(ctx)
	if err != nil {
		return fmt.Errorf("lifecycle: restore: list domains: %w", err)
	}
	// One component decides who starts, and it is this one: clear the
	// libvirt autostart flag again on every domain when the hypervisor
	// supports it. Create already clears it at define time.
	if clearer, ok := m.hyp.(AutostartClearer); ok {
		for _, dom := range domains {
			if err := clearer.ClearAutostart(ctx, dom.Name); err != nil {
				m.log.Warn("restore: autostart not cleared", "domain", dom.Name, "error", err)
			}
		}
	}
	states := make(map[string]types.State, len(domains))
	for _, dom := range domains {
		states[dom.UUID] = dom.State
	}
	if err := m.store.UpdateObservedStates(states); err != nil {
		return fmt.Errorf("lifecycle: restore: record observed states: %w", err)
	}

	insts, err := m.store.InstancesToRestore()
	if err != nil {
		return fmt.Errorf("lifecycle: restore: %w", err)
	}
	if len(insts) == 0 {
		m.log.Info("restore: nothing to start")
		return nil
	}
	batches := (len(insts) + m.batchSize - 1) / m.batchSize
	m.log.Info("restore: starting instances", "instances", len(insts), "batches", batches, "batch_size", m.batchSize)

	for i := 0; i < len(insts); i += m.batchSize {
		batch := insts[i:min(i+m.batchSize, len(insts))]
		m.log.Info("restore: batch starting", "batch", i/m.batchSize+1, "of", batches, "instances", len(batch))

		started := make([]types.Instance, 0, len(batch))
		for _, inst := range batch {
			if err := m.store.SetObservedState(inst.UUID, types.StateStarting); err != nil {
				m.log.Warn("restore: starting state not recorded", "instance", inst.Name, "error", err)
			}
			if err := m.hyp.Start(ctx, inst.Name); err != nil {
				m.log.Warn("restore: instance did not start", "instance", inst.Name, "error", err)
				m.syncObserved(ctx, inst)
				continue
			}
			started = append(started, inst)
		}
		for _, inst := range started {
			if err := m.waitRunning(ctx, inst.Name); err != nil {
				m.log.Warn("restore: instance did not reach running", "instance", inst.Name, "error", err)
				m.syncObserved(ctx, inst)
				continue
			}
			if err := m.store.SetObservedState(inst.UUID, types.StateRunning); err != nil {
				m.log.Warn("restore: running state not recorded", "instance", inst.Name, "error", err)
			}
			m.log.Info("restore: instance running", "instance", inst.Name)
		}
		m.log.Info("restore: batch done", "batch", i/m.batchSize+1, "of", batches)
	}
	m.log.Info("restore: complete")
	return nil
}

// waitRunning polls the observed state of one domain until it is running,
// up to the configured start timeout.
func (m *Manager) waitRunning(ctx context.Context, name string) error {
	attempts := int(m.startWait / m.startPoll)
	if attempts < 1 {
		attempts = 1
	}
	var last types.State
	for i := 0; i < attempts; i++ {
		if err := ctx.Err(); err != nil {
			return err
		}
		state, err := m.hyp.State(ctx, name)
		if err != nil {
			return err
		}
		last = state
		if state == types.StateRunning {
			return nil
		}
		if i < attempts-1 {
			m.sleep(m.startPoll)
		}
	}
	return fmt.Errorf("lifecycle: %s not running after %s (last state %s)", name, m.startWait, last)
}

// syncObserved records what libvirt reports for one instance, so a failed
// start never leaves "starting" stuck in the store.
func (m *Manager) syncObserved(ctx context.Context, inst types.Instance) {
	state, err := m.hyp.State(ctx, inst.Name)
	if err != nil {
		state = types.StateStopped
	}
	if err := m.store.SetObservedState(inst.UUID, state); err != nil {
		m.log.Warn("observed state not recorded", "instance", inst.Name, "error", err)
	}
}

File: crates/lifecycle/src/restore.rs
use std::collections::HashMap;

use bento_types::{Instance, State};

use crate::{Error, Manager, Result};

impl Manager {
    /// Restores each instance to its last desired state after host reboot
    /// (SPEC 11.2). Starts are batched and every batch reaches running or
    /// times out before the next, avoiding a simultaneous memory/disk spike.
    pub async fn restore(&self) -> Result<()> {
        let domains = self.hyp.list().await.map_err(|error| {
            Error::operation(format!("lifecycle: restore: list domains: {error}"))
        })?;
        if let Some(clearer) = &self.autostart_clearer {
            for domain in &domains {
                if let Err(error) = clearer.clear_autostart(&domain.name).await {
                    self.log.warn(&format!(
                        "restore: autostart not cleared for {}: {error}",
                        domain.name
                    ));
                }
            }
        } else {
            self.log.warn("restore: hypervisor cannot clear autostart; persistent definitions catch up when redefined");
        }
        let states: HashMap<_, _> = domains
            .into_iter()
            .map(|domain| (domain.uuid, domain.state))
            .collect();
        self.store
            .update_observed_states(states)
            .await
            .map_err(|error| {
                Error::operation(format!(
                    "lifecycle: restore: record observed states: {error}"
                ))
            })?;
        let instances = self
            .store
            .instances_to_restore()
            .await
            .map_err(|error| Error::operation(format!("lifecycle: restore: {error}")))?;
        if instances.is_empty() {
            self.log.info("restore: nothing to start");
            return Ok(());
        }
        let batches = instances.len().div_ceil(self.batch_size);
        self.log.info(&format!(
            "restore: starting {} instances in {batches} batches",
            instances.len()
        ));
        for (index, batch) in instances.chunks(self.batch_size).enumerate() {
            self.log
                .info(&format!("restore: batch {} starting", index + 1));
            let mut started = Vec::new();
            for instance in batch {
                if let Err(error) = self
                    .store
                    .set_observed_state(&instance.uuid, State::Starting)
                    .await
                {
                    self.log
                        .warn(&format!("restore: starting state not recorded: {error}"));
                }
                if let Err(error) = self.hyp.start(&instance.name).await {
                    self.log.warn(&format!(
                        "restore: {} did not start: {error}",
                        instance.name
                    ));
                    self.sync_observed(instance).await;
                } else {
                    started.push(instance);
                }
            }
            for instance in started {
                if let Err(error) = self.wait_running(&instance.name).await {
                    self.log.warn(&format!(
                        "restore: {} did not reach running: {error}",
                        instance.name
                    ));
                    self.sync_observed(instance).await;
                } else if let Err(error) = self
                    .store
                    .set_observed_state(&instance.uuid, State::Running)
                    .await
                {
                    self.log
                        .warn(&format!("restore: running state not recorded: {error}"));
                }
            }
            self.log.info(&format!("restore: batch {} done", index + 1));
        }
        self.log.info("restore: complete");
        Ok(())
    }

    async fn wait_running(&self, name: &str) -> Result<()> {
        let attempts = std::cmp::max(
            1,
            (self.start_wait.as_nanos() / self.start_poll.as_nanos()) as usize,
        );
        let mut last = State::Stopped;
        for attempt in 0..attempts {
            last = self.hyp.state(name).await.map_err(Error::caused)?;
            if last == State::Running {
                return Ok(());
            }
            if attempt + 1 < attempts {
                self.sleep.sleep(self.start_poll).await;
            }
        }
        Err(Error::operation(format!(
            "lifecycle: {name} not running after {:?} (last state {last})",
            self.start_wait
        )))
    }

    async fn sync_observed(&self, instance: &Instance) {
        let state = self
            .hyp
            .state(&instance.name)
            .await
            .unwrap_or(State::Stopped);
        if let Err(error) = self.store.set_observed_state(&instance.uuid, state).await {
            self.log
                .warn(&format!("observed state not recorded: {error}"));
        }
    }
}
